Rootless Docker support (`container init caused: write sysctl key net.ipv4.ping_group_range: write /proc/sys/net/ipv4/ping_group_range: invalid argument: unknown`)

Is there support for using buildx command with a rootless docker daemon? I tried the default buildkit image and the rootless one but both raises this error while booting up buildkit:
```
error: Error response from daemon: OCI runtime create failed: container_linux.go:370: starting container process caused: process_linux.go:459: container init caused: write sysctl key net.ipv4.ping_group_range: write /proc/sys/net/ipv4/ping_group_range: invalid argument: unknown
```
As a workaround, I found out that manually running buildkit with rootless [config](https://github.com/moby/buildkit/blob/master/docs/rootless.md) worked

```sh
docker run --name buildx_buildkit_builder0 -d --security-opt seccomp=unconfined --security-opt apparmor=unconfined --device /dev/fuse moby/buildkit:rootless --oci-worker-no-process-sandbox
```

You just gotta make sure that the name of the container matches the name of the buildx builder. In this case, my builder was named "builder" so the buildkit container had to be named "buildx_buildkit_builder0"

Here's logs from my most recent attempt, after updating everything I could.

```
$ docker info
Client:                                                                                                                  
 Context:    default                                        
 Debug Mode: false                                          
 Plugins:                                                   
  app: Docker App (Docker Inc., v0.9.1-beta3)               
  buildx: Docker Buildx (Docker Inc., v0.7.1-docker)
  scan: Docker Scan (Docker Inc., v0.12.0)                  
                                                            
Server:                                                     
 Containers: 0                
  Running: 0                                                
  Paused: 0                                                 
  Stopped: 0                                                
 Images: 6                                                  
 Server Version: 20.10.12                                   
 Storage Driver: overlay2    
  Backing Filesystem: extfs                                                                                              
  Supports d_type: true                                                                                                  
  Native Overlay Diff: false                                
  userxattr: true                                                                                                        
 Logging Driver: json-file                                                                                               
 Cgroup Driver: systemd                                                                                                  
 Cgroup Version: 2                                          
 Plugins:                                                   
  Volume: local                                             
  Network: bridge host ipvlan macvlan null overlay                                                                       
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog                                    
 Swarm: inactive                                                                                                         
 Runtimes: io.containerd.runc.v2 io.containerd.runtime.v1.linux runc                                                                                                        
 Default Runtime: runc                                      
 Init Binary: docker-init                                                                                                
 containerd version: 7b11cfaabd73bb80907dd23182b9347b4245eb5d                                                            
 runc version: v1.0.2-0-g52b36a2                            
 init version: de40ad0                                                                                                   
 Security Options:                                                                                                       
  seccomp                                                                                                                
   Profile: default                                                                                                      
  rootless                                                                                                               
  cgroupns                                                                                                               
 Kernel Version: 5.15.5-76051505-generic                                                                                 
 Operating System: Ubuntu 21.10                                                                                           
 OSType: linux                                              
 Architecture: x86_64                                       
 CPUs: 24                                                   
 Total Memory: 31.36GiB 
 Name: callisto                                                                                                          
 ID: SKP2:MDO2:TKSS:CC53:GS4T:DWSZ:JFUR:CTYU:NFVS:W67R:ELVO:235H                                                                                                            
 Docker Root Dir: /home/aaron/.local/share/docker                                     
 Debug Mode: false                                          
 Username: agable                                           
 Registry: https://index.docker.io/v1/                      
 Labels:                      
 Experimental: false                                        
 Insecure Registries:                                       
  127.0.0.0/8                                               
 Live Restore Enabled: false

$ docker buildx create --use --name=cross                   
cross

$ docker buildx build .                                                                                                  
WARN[0000] No output specified for docker-container driver. Build result will only remain in the build cache. To push result image into registry use --push or to load image
 into docker use --load                                                                                                  
[+] Building 1.5s (1/1) FINISHED                                                                                         
 => ERROR [internal] booting buildkit                                                                               1.5s 
 => => pulling image moby/buildkit:buildx-stable-1                                                                  1.1s                                                    
 => => creating container buildx_buildkit_cross0                                                                    0.4s                                                    
------                                                      
 > [internal] booting buildkit:                                                                                          
------                                                                                                                   
error: Error response from daemon: OCI runtime create failed: container_linux.go:380: starting container process caused: process_linux.go:545: container init caused: write sysctl key net.ipv4.ping_group_range: write /proc/sys/net/ipv4/ping_group_range: invalid argument: unknown
```

Unfortunately, it appears that I can't build docker from source on the master branch, as it uses buildx as part of its own build, and runs into the same error:
```
$ cd docker-ce
$ make static DOCKER_BUILD_PKGS=static-linux
make VERSION=20.10.0-beta1 CLI_DIR=/home/aaron/Code/docker-ce/components/cli ENGINE_DIR=/home/aaron/Code/docker-ce/components/engine -C /home/aaron/Code/docker-ce/components/packaging static
make[1]: Entering directory '/home/aaron/Code/docker-ce/components/packaging'
mkdir -p "src/github.com/docker"
cp -r "/home/aaron/Code/docker-ce/components/cli" src/github.com/docker/cli
./scripts/checkout.sh src/github.com/docker/cli "HEAD"
+ SRC=src/github.com/docker/cli
+ REF=HEAD
+ REF_FETCH=HEAD
+ git -C src/github.com/docker/cli ls-remote --refs --heads --tags origin HEAD
+ awk {print $2}
+ REF=
+ [ -n  ]
+ REF=FETCH_HEAD
+ git -C src/github.com/docker/cli fetch --update-head-ok --depth 1 origin HEAD
remote: Total 0 (delta 0), reused 0 (delta 0), pack-reused 0
From ssh://github.com/docker/docker-ce
 * branch                  HEAD       -> FETCH_HEAD
+ git -C src/github.com/docker/cli checkout -q FETCH_HEAD
mkdir -p "src/github.com/docker"
cp -r "/home/aaron/Code/docker-ce/components/engine" src/github.com/docker/docker
./scripts/checkout.sh src/github.com/docker/docker "HEAD"
+ SRC=src/github.com/docker/docker
+ REF=HEAD
+ REF_FETCH=HEAD
+ git -C src/github.com/docker/docker ls-remote --refs --heads --tags origin HEAD
+ awk {print $2}
+ REF=
+ [ -n  ]
+ REF=FETCH_HEAD
+ git -C src/github.com/docker/docker fetch --update-head-ok --depth 1 origin HEAD
remote: Total 0 (delta 0), reused 0 (delta 0), pack-reused 0
From ssh://github.com/docker/docker-ce
 * branch                  HEAD       -> FETCH_HEAD
+ git -C src/github.com/docker/docker checkout -q FETCH_HEAD
git init src/github.com/docker/scan-cli-plugin
Initialized empty Git repository in /home/aaron/Code/docker-ce/components/packaging/src/github.com/docker/scan-cli-plugin/.git/
git -C src/github.com/docker/scan-cli-plugin remote add origin "https://github.com/docker/scan-cli-plugin.git"
./scripts/checkout.sh src/github.com/docker/scan-cli-plugin "v0.10.0"
+ SRC=src/github.com/docker/scan-cli-plugin
+ REF=v0.10.0
+ REF_FETCH=v0.10.0
+ git -C src/github.com/docker/scan-cli-plugin ls-remote --refs --heads --tags origin v0.10.0
+ awk {print $2}
+ REF=refs/tags/v0.10.0
+ [ -n refs/tags/v0.10.0 ]
+ REF_FETCH=refs/tags/v0.10.0:refs/tags/v0.10.0
+ git -C src/github.com/docker/scan-cli-plugin fetch --update-head-ok --depth 1 origin refs/tags/v0.10.0:refs/tags/v0.10.0
remote: Enumerating objects: 71, done.
remote: Counting objects: 100% (71/71), done.
remote: Compressing objects: 100% (64/64), done.
remote: Total 71 (delta 8), reused 30 (delta 1), pack-reused 0
Unpacking objects: 100% (71/71), 111.10 KiB | 669.00 KiB/s, done.
From https://github.com/docker/scan-cli-plugin
 * [new tag]         v0.10.0    -> v0.10.0
+ git -C src/github.com/docker/scan-cli-plugin checkout -q refs/tags/v0.10.0
for p in static-linux; do \
        make -C static VERSION=20.10.0-beta1 GO_VERSION=1.16.9 TARGETPLATFORM= CONTAINERD_VERSION= RUNC_VERSION= ${p}; \
done
make[2]: Entering directory '/home/aaron/Code/docker-ce/components/packaging/static'
cd /home/aaron/Code/docker-ce/components/cli && VERSION=20.10.0-beta1 docker buildx bake --set binary.platform= --set binary.args.CGO_ENABLED= binary
[+] Building 1.6s (1/1) FINISHED                                                                                                                                            
 => ERROR [internal] booting buildkit                                                                                                                                  1.6s
 => => pulling image moby/buildkit:buildx-stable-1                                                                                                                     1.1s
 => => creating container buildx_buildkit_cross0                                                                                                                       0.5s
------
 > [internal] booting buildkit:
------
error: Error response from daemon: OCI runtime create failed: container_linux.go:380: starting container process caused: process_linux.go:545: container init caused: write sysctl key net.ipv4.ping_group_range: write /proc/sys/net/ipv4/ping_group_range: invalid argument: unknown
make[2]: *** [Makefile:97: static-cli] Error 1
make[2]: Leaving directory '/home/aaron/Code/docker-ce/components/packaging/static'
make[1]: *** [Makefile:81: static] Error 2
make[1]: Leaving directory '/home/aaron/Code/docker-ce/components/packaging'
make: *** [Makefile:30: static] Error 2
```

--- a/driver/docker-container/driver.go
+++ b/driver/docker-container/driver.go
@@ -37,6 +37,7 @@ const (
 type Driver struct {
 	driver.InitConfig
 	factory      driver.Factory
+	userNSRemap  bool // true if dockerd is running with userns-remap mode
 	netMode      string
 	image        string
 	cgroupParent string
@@ -112,7 +113,6 @@ func (d *Driver) create(ctx context.Context, l progress.SubLogger) error {
 	if err := l.Wrap("creating container "+d.Name, func() error {
 		hc := &container.HostConfig{
 			Privileged: true,
-			UsernsMode: "host",
 			Mounts: []mount.Mount{
 				{
 					Type:   mount.TypeVolume,
@@ -120,6 +120,9 @@ func (d *Driver) create(ctx context.Context, l progress.SubLogger) error {
 					Target: confutil.DefaultBuildKitStateDir,
 				},
 			},
+		}
+		if d.userNSRemap {
+			hc.UsernsMode = "host"
 		}
 		if d.netMode != "" {
 			hc.NetworkMode = container.NetworkMode(d.netMode)

--- a/driver/docker-container/factory.go
+++ b/driver/docker-container/factory.go
@@ -6,6 +6,7 @@ import (
 	"strings"
 
 	"github.com/docker/buildx/driver"
+	dockertypes "github.com/docker/docker/api/types"
 	dockerclient "github.com/docker/docker/client"
 	"github.com/pkg/errors"
 )
@@ -40,6 +41,20 @@ func (f *factory) New(ctx context.Context, cfg driver.InitConfig) (driver.Driver
 		return nil, errors.Errorf("%s driver requires docker API access", f.Name())
 	}
 	d := &Driver{factory: f, InitConfig: cfg}
+	dockerInfo, err := cfg.DockerAPI.Info(ctx)
+	if err != nil {
+		return nil, err
+	}
+	secOpts, err := dockertypes.DecodeSecurityOptions(dockerInfo.SecurityOptions)
+	if err != nil {
+		return nil, err
+	}
+	for _, f := range secOpts {
+		if f.Name == "userns" {
+			d.userNSRemap = true
+			break
+		}
+	}
 	for k, v := range cfg.DriverOpts {
 		switch {
 		case k == "network":

